File: src/main/java/resonantinduction/electrical/transformer/RenderTransformer.java
package resonantinduction.electrical.transformer;

import net.minecraft.util.ResourceLocation;
import net.minecraftforge.client.model.AdvancedModelLoader;
import net.minecraftforge.client.model.obj.WavefrontObject;

import org.lwjgl.opengl.GL11;

import resonantinduction.core.Reference;
import calclavia.lib.render.RenderUtility;
import cpw.mods.fml.client.FMLClientHandler;
import cpw.mods.fml.relauncher.Side;
import cpw.mods.fml.relauncher.SideOnly;

@SideOnly(Side.CLIENT)
public class RenderTransformer
{
	public static final RenderTransformer INSTANCE = new RenderTransformer();

	public static final WavefrontObject MODEL = (WavefrontObject) AdvancedModelLoader.loadModel(Reference.MODEL_DIRECTORY + "transformer.obj");
	public static final ResourceLocation TEXTURE_COIL = new ResourceLocation(Reference.DOMAIN, Reference.MODEL_PATH + "transformer_winding.png");
	public static final ResourceLocation TEXTURE_STONE = new ResourceLocation(Reference.BLOCK_TEXTURE_DIRECTORY + "stone.png");
	public static final ResourceLocation TEXTURE_IRON = new ResourceLocation(Reference.BLOCK_TEXTURE_DIRECTORY + "iron_block.png");

	public void doRender()
	{
		GL11.glScalef(0.5f, 0.5f, 0.5f);
		FMLClientHandler.instance().getClient().renderEngine.bindTexture(TEXTURE_COIL);
		MODEL.renderAllExcept("core", "base");
		FMLClientHandler.instance().getClient().renderEngine.bindTexture(TEXTURE_IRON);
		MODEL.renderOnly("core");
		FMLClientHandler.instance().getClient().renderEngine.bindTexture(TEXTURE_STONE);
		MODEL.renderOnly("base");
	}

	public void render(PartTransformer part, double x, double y, double z)
	{
		GL11.glPushMatrix();
		GL11.glTranslatef((float) x + 0.5F, (float) y + 0.5F, (float) z + 0.5F);
		RenderUtility.rotateFaceBlockToSide(part.placementSide);
		GL11.glRotatef(90, 0, 1, 0);
		RenderUtility.rotateBlockBasedOnDirection(part.getFacing());
		GL11.glScalef(0.5f, 0.5f, 0.5f);
		FMLClientHandler.instance().getClient().renderEngine.bindTexture(TEXTURE_COIL);

		switch (part.multiplier)
		{
			case 0:
				MODEL.renderOnly("InsulatorLayerLow", "OuterWindingLowBox", "InnerWindingLowBox");
				break;
			case 1:
				MODEL.renderOnly("InsulatorLayerMed", "OuterWindingMedBox", "InnerWindingMedBox");
				break;
			case 2:
				MODEL.renderOnly("InnerWindingHighBox", "InsulatorLayerHigh", "OuterWindingHighBox");
				break;
		}

		MODEL.renderOnly("OuterWindingConnector", "InnerWindingConnector");

		FMLClientHandler.instance().getClient().renderEngine.bindTexture(TEXTURE_IRON);
		MODEL.renderOnly("core");
		FMLClientHandler.instance().getClient().renderEngine.bindTexture(TEXTURE_STONE);
		MODEL.renderOnly("base");
		GL11.glPopMatrix();
	}
}

File: src/main/scala/resonantinduction/electrical/transformer/RenderTransformer.scala
package resonantinduction.electrical.transformer

import cpw.mods.fml.client.FMLClientHandler
import cpw.mods.fml.relauncher.Side
import cpw.mods.fml.relauncher.SideOnly
import net.minecraft.item.ItemStack
import net.minecraft.util.ResourceLocation
import net.minecraftforge.client.IItemRenderer
import net.minecraftforge.client.model.AdvancedModelLoader
import net.minecraftforge.client.model.obj.WavefrontObject
import org.lwjgl.opengl.GL11
import resonant.content.prefab.scal.render.ISimpleItemRenderer
import resonant.lib.render.RenderUtility
import resonantinduction.core.Reference

@SideOnly(Side.CLIENT) object RenderTransformer
{
    final val INSTANCE: RenderTransformer = new RenderTransformer
    final val MODEL: WavefrontObject = AdvancedModelLoader.loadModel(new ResourceLocation(Reference.domain, Reference.modelPath + "transformer.obj")).asInstanceOf[WavefrontObject]
    final val TEXTURE_COIL: ResourceLocation = new ResourceLocation(Reference.domain, Reference.modelPath + "transformer_winding.png")
    final val TEXTURE_STONE: ResourceLocation = new ResourceLocation(Reference.blockTextureDirectory + "stone.png")
    final val TEXTURE_IRON: ResourceLocation = new ResourceLocation(Reference.blockTextureDirectory + "iron_block.png")
}

@SideOnly(Side.CLIENT) class RenderTransformer extends ISimpleItemRenderer
{
    def renderInventoryItem(`type`: IItemRenderer.ItemRenderType, itemStack: ItemStack, data: AnyRef*)
    {
        GL11.glTranslated(0, -0.2f, 0)
        doRender
    }

    def doRender
    {
        GL11.glScalef(0.5f, 0.5f, 0.5f)
        FMLClientHandler.instance.getClient.renderEngine.bindTexture(RenderTransformer.TEXTURE_COIL)
        RenderTransformer.MODEL.renderAllExcept("core", "base")
        FMLClientHandler.instance.getClient.renderEngine.bindTexture(RenderTransformer.TEXTURE_IRON)
        RenderTransformer.MODEL.renderOnly("core")
        FMLClientHandler.instance.getClient.renderEngine.bindTexture(RenderTransformer.TEXTURE_STONE)
        RenderTransformer.MODEL.renderOnly("base")
    }

    def render(part: PartElectricTransformer, x: Double, y: Double, z: Double)
    {
        GL11.glPushMatrix
        GL11.glTranslatef(x.asInstanceOf[Float] + 0.5F, y.asInstanceOf[Float] + 0.5F, z.asInstanceOf[Float] + 0.5F)
        RenderUtility.rotateFaceBlockToSide(part.placementSide)
        RenderUtility.rotateBlockBasedOnDirection(part.getFacing)
        GL11.glRotatef(90, 0, 1, 0)
        GL11.glScalef(0.5f, 0.5f, 0.5f)
        FMLClientHandler.instance.getClient.renderEngine.bindTexture(RenderTransformer.TEXTURE_COIL)
        if (part.multiplier == 0)
        {
            RenderTransformer.MODEL.renderOnly("InsulatorLayerLow", "OuterWindingLowBox", "InnerWindingLowBox")
        }
        else if (part.multiplier == 1)
        {
            RenderTransformer.MODEL.renderOnly("InsulatorLayerMed", "OuterWindingMedBox", "InnerWindingMedBox")
        }
        else if (part.multiplier == 2)
        {
            RenderTransformer.MODEL.renderOnly("InnerWindingHighBox", "InsulatorLayerHigh", "OuterWindingHighBox")
        }
        RenderTransformer.MODEL.renderOnly("OuterWindingConnector", "InnerWindingConnector")
        FMLClientHandler.instance.getClient.renderEngine.bindTexture(RenderTransformer.TEXTURE_IRON)
        RenderTransformer.MODEL.renderOnly("core")
        FMLClientHandler.instance.getClient.renderEngine.bindTexture(RenderTransformer.TEXTURE_STONE)
        RenderTransformer.MODEL.renderOnly("base")
        GL11.glPopMatrix
    }
}